### PR TITLE
install: suppress redundunt warnings with `depends_on` requirement

### PR DESCRIPTION
**Pull request by umireon**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

When a formula depends on any [requirements], they evaluated at most three times.

- Before locking: [verify_deps_exist in FormulaInstaller#prelude]
- Before installing: [compute_dependencies in FormulaInstaller#install]
- Before building formula: [compute_and_install_dependencies in FormulaInstaller#install]

These methods call [FormulaInstaller#check_requirements], which `puts` the warning messages generated by the specified requirement class.

The problem is when a non-fatal requirement is specified and thus evaluated three times,
mostly the same warning message is also emitted three times.

[requirements]: https://github.com/Homebrew/brew/blob/59e2d6772158d8df0735708ae78ddec6ccc68026/Library/Homebrew/requirement.rb
[verify_deps_exist in FormulaInstaller#prelude]: https://github.com/Homebrew/brew/blob/59e2d6772158d8df0735708ae78ddec6ccc68026/Library/Homebrew/formula_installer.rb#L127
[compute_dependencies in FormulaInstaller#install]: https://github.com/Homebrew/brew/blob/59e2d6772158d8df0735708ae78ddec6ccc68026/Library/Homebrew/formula_installer.rb#L200
[compute_and_install_dependencies in FormulaInstaller#install]: https://github.com/Homebrew/brew/blob/59e2d6772158d8df0735708ae78ddec6ccc68026/Library/Homebrew/formula_installer.rb#L262
[FormulaInstaller#check_requirements]: https://github.com/Homebrew/brew/blob/59e2d6772158d8df0735708ae78ddec6ccc68026/Library/Homebrew/formula_installer.rb#L336-L348

## Example
```
$ brew install python
$ brew install gdb --with-python
Updating Homebrew...
gdb: A build of GDB using a brewed Python was requested, but Python is not
a universal build.

GDB requires Python to be built as a universal binary or it will fail
if attempting to debug a 32-bit binary on a 64-bit host.
gdb: A build of GDB using a brewed Python was requested, but Python is not
a universal build.

GDB requires Python to be built as a universal binary or it will fail
if attempting to debug a 32-bit binary on a 64-bit host.
gdb: A build of GDB using a brewed Python was requested, but Python is not
a universal build.

GDB requires Python to be built as a universal binary or it will fail
if attempting to debug a 32-bit binary on a 64-bit host.
==> Using the sandbox
==> Downloading https://ftpmirror.gnu.org/gdb/gdb-7.12.tar.xz

```

This change restricts printing the warning messages only when a bottle is
successfully installed or before building.
Since this timing is after the final dependency computation for each cases,
the warnings will be most useful to check what is not yet satisfied.

### CODE REVIEW

File: Library/Homebrew/formula_installer.rb [hunk 1]
```diff
@@ -251,6 +251,7 @@ def install
         opoo "Bottle installation failed: building from source."
         raise BuildToolsError, [formula] unless DevelopmentTools.installed?
       else
+        puts @requirement_messages
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb]: Make a new function with `puts @requirement_messages` like `def puts_requirement_messages` or similar which outputs nothing i.e. no newline which `@requirement_messages` is empty

Comment (umireon) [inline Library/Homebrew/formula_installer.rb]: `@requirement_messages` is an `Array` and AFAIK `puts []` outputs nothing (no newline).
Therefore, this code already works as you suggested.

**Comment by umireon** [comment]
Thank you for reviewing, @MikeMcQuaid !
IMO `puts_requirement_messages` is too trivial to define, but it makes senses in encapsulation.
```ruby
def puts_requirement_messages
  puts @requirement_messages
end
```

**Comment by umireon** [comment]
Not relevant to this PR, however, the actual problem might be evaluating the requirements three times.
When I was modifying a Formula with Requirement class, I expected the requirement evaluated once, but it was not the case.

**Comment by MikeMcQuaid** [comment]
> IMO puts_requirement_messages is too trivial to define, but it makes senses in encapsulation.

As mentioned there's more logic for it to do: don't `puts` anything if it's empty.

**Comment by MikeMcQuaid** [comment]
> As mentioned there's more logic for it to do: don't puts anything if it's empty.

So something like:

```ruby
def puts_requirement_messages
  return unless @requirement_messages
  return if @requirement_messages.empty?
  puts @requirement_messages
end
```

**Comment by umireon** [comment]
Add puts_requirement_messages for the case `@requirement_messages` is a String or nil!

**Comment by umireon** [comment]
`@requirement_messages` should be initialized with an empty string (`""`) instead of an empty Array (`[]`)?

**Comment by vladshablinsky** [comment]
> @requirement_messages should be initialized with an empty string ("") instead of an empty Array ([])?

I think everything is fine with current initialization.

**Comment by umireon** [comment]
I think the suggested guards `return unless @requirement_messages; return if @requirement_messages.empty?` are unnecessary when `@requirement_messages` is initialized with `[]`.

**Comment by MikeMcQuaid** [comment]
> are unnecessary when @requirement_messages is initialized with [].

If it's called before initialisation in future then it will output an empty line if it's `nil` which is what we want to avoid.

**Comment by umireon** [comment]
Anyway, the foolproof version of `puts_requirement_messages` is implemented in bcc9002 .
I understand this kind of robustness is important for such a big OSS product.
Thank you for discussing, @MikeMcQuaid , @vladshablinsky 🍎

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request merged**